<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Quick Start &mdash; phykit  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=b745eb55" />

  
    <link rel="shortcut icon" href="_static/flavicon.png"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=5929fcd5"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="About" href="about/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#">
            
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage/index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="change_log/index.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="other_software/index.html">Other software</a></li>
<li class="toctree-l1"><a class="reference internal" href="frequently_asked_questions/index.html">FAQ</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">phykit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Quick Start</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <a class="reference external image-reference" href="https://jlsteenwyk.com/PhyKIT"><img alt="_images/logo.png" class="align-center" src="_images/logo.png" style="width: 55%;" /></a>
<hr class="docutils" />
<p>PhyKIT, a toolkit for the UNIX shell environment with numerous functions that process multiple
sequence alignments and phylogenies for broad applications</p>
<p>If you found PhyKIT useful, please cite <em>PhyKIT: a broadly applicable UNIX shell toolkit for processing
and analyzing phylogenomic data</em>. Bioinformatics. doi: <a class="reference external" href="https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btab096/6131675?redirectedFrom=fulltext">10.1093/bioinformatics/btab096</a>.</p>
<section id="quick-start">
<h1>Quick Start<a class="headerlink" href="#quick-start" title="Link to this heading">ÔÉÅ</a></h1>
<p>These two lines represent the simpliest method to rapidly install and run PhyKIT.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1"># install</span>
pip<span class="w"> </span>install<span class="w"> </span>phykit
<span class="c1"># run</span>
phykit<span class="w"> </span>-h
</pre></div>
</div>
<p><strong>1) Installation</strong></p>
<p>To install using <em>pip</em>, we strongly recommend building a virtual environment to avoid
software dependency issues. To do so, execute the following commands:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1"># create virtual environment</span>
python<span class="w"> </span>-m<span class="w"> </span>venv<span class="w"> </span>venv
<span class="c1"># activate virtual environment</span>
<span class="nb">source</span><span class="w"> </span>venv/bin/activate
<span class="c1"># install phykit</span>
pip<span class="w"> </span>install<span class="w"> </span>phykit
</pre></div>
</div>
<p><strong>Note, the virtual environment must be activated to use phykit.</strong></p>
<p>After using PhyKIT, you may wish to deactivate your virtual environment and can do so using the following command:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1"># deactivate virtual environment</span>
deactivate
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p>Similarly, to install from source, we strongly recommend using a virtual environment. To do so, use the
following commands:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1"># download</span>
git<span class="w"> </span>clone<span class="w"> </span>https://github.com/JLSteenwyk/PhyKIT.git
<span class="nb">cd</span><span class="w"> </span>PhyKIT/
<span class="c1"># create virtual environment</span>
python<span class="w"> </span>-m<span class="w"> </span>venv<span class="w"> </span>venv
<span class="c1"># activate virtual environment</span>
<span class="nb">source</span><span class="w"> </span>venv/bin/activate
<span class="c1"># install</span>
make<span class="w"> </span>install
</pre></div>
</div>
<p>To deactivate your virtual environment, use the following command:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1"># deactivate virtual environment</span>
deactivate
</pre></div>
</div>
<p><strong>Note, the virtual environment must be activated to use phykit.</strong></p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<p>To install via anaconda, execute the following command:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>conda<span class="w"> </span>install<span class="w"> </span>bioconda::phykit
</pre></div>
</div>
<p>Visit here for more information:
<a class="reference external" href="https://anaconda.org/bioconda/phykit">https://anaconda.org/bioconda/phykit</a></p>
<p><strong>2) Usage</strong></p>
<p>Get the help message from PhyKIT:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>-h
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
<hr class="docutils" />
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="about/index.html">About</a><ul>
<li class="toctree-l2"><a class="reference internal" href="about/index.html#the-developers">The Developers</a></li>
<li class="toctree-l2"><a class="reference internal" href="about/index.html#more-team-members">More Team Members</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usage/index.html">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="usage/index.html#general-usage">General usage</a><ul>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#calling-functions">Calling functions</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#function-aliases">Function aliases</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#command-line-interfaces">Command line interfaces</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="usage/index.html#alignment-based-functions">Alignment-based functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#alignment-length">Alignment length</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#alignment-length-no-gaps">Alignment length no gaps</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#alignment-recoding">Alignment recoding</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#column-score">Column score</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#compositional-bias-per-site">Compositional bias per site</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#create-concatenation-matrix">Create concatenation matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#evolutionary-rate-per-site">Evolutionary Rate per Site</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#faidx">Faidx</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#guanine-cytosine-gc-content">Guanine-cytosine (GC) content</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#pairwise-identity">Pairwise identity</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#parsimony-informative-sites">Parsimony informative sites</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#protein-to-nucleotide-alignment">Protein-to-nucleotide alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#relative-composition-variability">Relative composition variability</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#relative-composition-variability-taxon">Relative composition variability, taxon</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#rename-fasta-entries">Rename FASTA entries</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#sum-of-pairs-score">Sum-of-pairs score</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#variable-sites">Variable sites</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="usage/index.html#tree-based-functions">Tree-based functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#bipartition-support-statistics">Bipartition support statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#branch-length-multiplier">Branch length multiplier</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#collapse-bipartitions">Collapse bipartitions</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#covarying-evolutionary-rates">Covarying evolutionary rates</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#degree-of-violation-of-the-molecular-clock">Degree of violation of the molecular clock</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#evolutionary-rate">Evolutionary rate</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#hidden-paralogy-check">Hidden paralogy check</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#internal-branch-statistics">Internal branch statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#internode-labeler">Internode labeler</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#last-common-ancestor-subtree">Last common ancestor subtree</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#long-branch-score">Long branch score</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#monophyly-check">Monophyly check</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#nearest-neighbor-interchange">Nearest neighbor interchange</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#patristic-distances">Patristic distances</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#polytomy-testing">Polytomy testing</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#print-tree">Print tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#prune-tree">Prune tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#rename-tree-tips">Rename tree tips</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#robinson-foulds-distance">Robinson-Foulds distance</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#root-tree">Root tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#spurious-homolog-identification">Spurious homolog identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#terminal-branch-statistics">Terminal branch statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#tip-labels">Tip labels</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#tip-to-tip-distance">Tip-to-tip distance</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#tip-to-tip-node-distance">Tip-to-tip node distance</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#total-tree-length">Total tree length</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#treeness">Treeness</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="usage/index.html#alignment-and-tree-based-functions">Alignment- and tree-based functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#saturation">Saturation</a></li>
<li class="toctree-l3"><a class="reference internal" href="usage/index.html#treeness-over-rcv">Treeness over RCV</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/index.html">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="tutorials/index.html#summarizing-information-content">1. Summarizing information content</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#alignment-length">Alignment length</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#bipartition-support-statistics">Bipartition support statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#long-branch-score">Long branch score</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#parsimony-informative-sites">Parsimony informative sites</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#saturation">Saturation</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#treeness-divided-by-relative-composition-variability">Treeness divided by relative composition variability</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#variable-sites">Variable sites</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/index.html#evaluating-gene-gene-covariation">2. Evaluating gene-gene covariation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#step-0-prepare-data">Step 0: Prepare data</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#step-1-estimate-gene-tree-branch-lengths">Step 1: Estimate gene tree branch lengths</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#step-2-root-constrain-trees">Step 2: Root constrain trees</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#step-3-evaluate-gene-gene-covariation">Step 3: Evaluate gene-gene covariation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/index.html#identifying-signatures-of-rapid-radiations">3. Identifying signatures of rapid radiations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#id16">Step 0: Prepare data</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#step-1-conduct-polytomy-test">Step 1: Conduct polytomy test</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tutorials/index.html#evaluating-the-accuracy-of-a-multiple-sequence-alignment">4. Evaluating the accuracy of a multiple sequence alignment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#step-0-generate-query-alignments">Step 0: Generate query alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="tutorials/index.html#step-1-score-each-alignment">Step 1: Score each alignment</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="change_log/index.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="other_software/index.html">Other software</a></li>
<li class="toctree-l1"><a class="reference internal" href="frequently_asked_questions/index.html">FAQ</a></li>
</ul>
</div>
<hr class="docutils" />
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="about/index.html" class="btn btn-neutral float-right" title="About" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020 Jacob L. Steenwyk.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-104875636-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-104875636-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>