<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; phykit  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=b745eb55" />

  
    <link rel="shortcut icon" href="../_static/flavicon.png"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tutorials" href="../tutorials/index.html" />
    <link rel="prev" title="About" href="../about/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../about/index.html">About</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#general-usage">General usage</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#calling-functions">Calling functions</a></li>
<li class="toctree-l3"><a class="reference internal" href="#function-aliases">Function aliases</a></li>
<li class="toctree-l3"><a class="reference internal" href="#command-line-interfaces">Command line interfaces</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#alignment-based-functions">Alignment-based functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#alignment-length">Alignment length</a></li>
<li class="toctree-l3"><a class="reference internal" href="#alignment-length-no-gaps">Alignment length no gaps</a></li>
<li class="toctree-l3"><a class="reference internal" href="#alignment-recoding">Alignment recoding</a></li>
<li class="toctree-l3"><a class="reference internal" href="#column-score">Column score</a></li>
<li class="toctree-l3"><a class="reference internal" href="#compositional-bias-per-site">Compositional bias per site</a></li>
<li class="toctree-l3"><a class="reference internal" href="#create-concatenation-matrix">Create concatenation matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="#evolutionary-rate-per-site">Evolutionary Rate per Site</a></li>
<li class="toctree-l3"><a class="reference internal" href="#faidx">Faidx</a></li>
<li class="toctree-l3"><a class="reference internal" href="#guanine-cytosine-gc-content">Guanine-cytosine (GC) content</a></li>
<li class="toctree-l3"><a class="reference internal" href="#pairwise-identity">Pairwise identity</a></li>
<li class="toctree-l3"><a class="reference internal" href="#parsimony-informative-sites">Parsimony informative sites</a></li>
<li class="toctree-l3"><a class="reference internal" href="#protein-to-nucleotide-alignment">Protein-to-nucleotide alignment</a></li>
<li class="toctree-l3"><a class="reference internal" href="#relative-composition-variability">Relative composition variability</a></li>
<li class="toctree-l3"><a class="reference internal" href="#relative-composition-variability-taxon">Relative composition variability, taxon</a></li>
<li class="toctree-l3"><a class="reference internal" href="#rename-fasta-entries">Rename FASTA entries</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sum-of-pairs-score">Sum-of-pairs score</a></li>
<li class="toctree-l3"><a class="reference internal" href="#variable-sites">Variable sites</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#tree-based-functions">Tree-based functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#bipartition-support-statistics">Bipartition support statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#branch-length-multiplier">Branch length multiplier</a></li>
<li class="toctree-l3"><a class="reference internal" href="#collapse-bipartitions">Collapse bipartitions</a></li>
<li class="toctree-l3"><a class="reference internal" href="#covarying-evolutionary-rates">Covarying evolutionary rates</a></li>
<li class="toctree-l3"><a class="reference internal" href="#degree-of-violation-of-the-molecular-clock">Degree of violation of the molecular clock</a></li>
<li class="toctree-l3"><a class="reference internal" href="#evolutionary-rate">Evolutionary rate</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hidden-paralogy-check">Hidden paralogy check</a></li>
<li class="toctree-l3"><a class="reference internal" href="#internal-branch-statistics">Internal branch statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#internode-labeler">Internode labeler</a></li>
<li class="toctree-l3"><a class="reference internal" href="#last-common-ancestor-subtree">Last common ancestor subtree</a></li>
<li class="toctree-l3"><a class="reference internal" href="#long-branch-score">Long branch score</a></li>
<li class="toctree-l3"><a class="reference internal" href="#monophyly-check">Monophyly check</a></li>
<li class="toctree-l3"><a class="reference internal" href="#nearest-neighbor-interchange">Nearest neighbor interchange</a></li>
<li class="toctree-l3"><a class="reference internal" href="#patristic-distances">Patristic distances</a></li>
<li class="toctree-l3"><a class="reference internal" href="#polytomy-testing">Polytomy testing</a></li>
<li class="toctree-l3"><a class="reference internal" href="#print-tree">Print tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="#prune-tree">Prune tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="#rename-tree-tips">Rename tree tips</a></li>
<li class="toctree-l3"><a class="reference internal" href="#robinson-foulds-distance">Robinson-Foulds distance</a></li>
<li class="toctree-l3"><a class="reference internal" href="#root-tree">Root tree</a></li>
<li class="toctree-l3"><a class="reference internal" href="#spurious-homolog-identification">Spurious homolog identification</a></li>
<li class="toctree-l3"><a class="reference internal" href="#terminal-branch-statistics">Terminal branch statistics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tip-labels">Tip labels</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tip-to-tip-distance">Tip-to-tip distance</a></li>
<li class="toctree-l3"><a class="reference internal" href="#tip-to-tip-node-distance">Tip-to-tip node distance</a></li>
<li class="toctree-l3"><a class="reference internal" href="#total-tree-length">Total tree length</a></li>
<li class="toctree-l3"><a class="reference internal" href="#treeness">Treeness</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#alignment-and-tree-based-functions">Alignment- and tree-based functions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#saturation">Saturation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#treeness-over-rcv">Treeness over RCV</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../change_log/index.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../other_software/index.html">Other software</a></li>
<li class="toctree-l1"><a class="reference internal" href="../frequently_asked_questions/index.html">FAQ</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">phykit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Usage</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="usage">
<span id="id1"></span><h1>Usage<a class="headerlink" href="#usage" title="Link to this heading">ÔÉÅ</a></h1>
<p>PhyKIT helps process and analyze multiple sequence alignments and phylogenies.</p>
<p>Generally, all functions are designed to help understand the contents of alignments
(e.g., gc content or the number of parsimony informative sites) and the shape
of trees (e.g., treeness, degree of violation of a molecular clock).</p>
<p>Some help messages indicate that summary statistics are reported (e.g.,
bipartition_support_stats). Summary statistics include mean, median, 25th percentile,
75th percentile, minimum, maximum, standard deviation, and variance. These functions
typically have a verbose option that allows users to get the underlying data
used to calculate summary statistics.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="general-usage">
<h2>General usage<a class="headerlink" href="#general-usage" title="Link to this heading">ÔÉÅ</a></h2>
<hr class="docutils" />
<section id="calling-functions">
<h3>Calling functions<a class="headerlink" href="#calling-functions" title="Link to this heading">ÔÉÅ</a></h3>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>&lt;command&gt;<span class="w"> </span><span class="o">[</span>optional<span class="w"> </span><span class="nb">command</span><span class="w"> </span>arguments<span class="o">]</span>
</pre></div>
</div>
<p>Command specific help messages can be viewed by adding a
-h/\-\-help argument after the command. For example, to see the help message
for the command ‚Äòtreeness‚Äô, execute:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>treeness<span class="w"> </span>-h
<span class="c1"># or</span>
phykit<span class="w"> </span>treeness<span class="w"> </span>--help
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="function-aliases">
<h3>Function aliases<a class="headerlink" href="#function-aliases" title="Link to this heading">ÔÉÅ</a></h3>
<p>Each function comes with aliases to save the user some
key strokes. For example, to get the help message for the ‚Äòtreeness‚Äô
function, you can type:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>tness<span class="w"> </span>-h
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="command-line-interfaces">
<h3>Command line interfaces<a class="headerlink" href="#command-line-interfaces" title="Link to this heading">ÔÉÅ</a></h3>
<p>As of version 1.2.0, all functions (including aliases) can be executed using
a command line interface that starts with <em>pk_</em>. For example, instead of typing
the previous command to get the help message of the treeness function, you can type:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>pk_treeness<span class="w"> </span>-h
<span class="c1"># or</span>
pk_tness<span class="w"> </span>-h
</pre></div>
</div>
<p>All possible function names are specified at the top of each function section.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="alignment-based-functions">
<h2>Alignment-based functions<a class="headerlink" href="#alignment-based-functions" title="Link to this heading">ÔÉÅ</a></h2>
<hr class="docutils" />
<section id="alignment-length">
<h3>Alignment length<a class="headerlink" href="#alignment-length" title="Link to this heading">ÔÉÅ</a></h3>
<a class="reference internal image-reference" href="../_images/aln_len.png"><img alt="../_images/aln_len.png" class="align-center" src="../_images/aln_len.png" style="width: 75%;" /></a>
<p>Function names: alignment_length; aln_len; al <br/>
Command line interface: pk_alignment_length; pk_aln_len; pk_al</p>
<p>Length of an input alignment is calculated using this function.</p>
<p>Longer alignments are associated with strong phylogenetic signal.</p>
<p>Association between alignment length and phylogenetic signal
was determined by Shen et al., Genome Biology and Evolution (2016),
doi: 10.1093/gbe/evw179.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>aln_len<span class="w"> </span>&lt;alignment&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="alignment-length-no-gaps">
<h3>Alignment length no gaps<a class="headerlink" href="#alignment-length-no-gaps" title="Link to this heading">ÔÉÅ</a></h3>
<a class="reference internal image-reference" href="../_images/aln_len_no_gaps.png"><img alt="../_images/aln_len_no_gaps.png" class="align-center" src="../_images/aln_len_no_gaps.png" style="width: 75%;" /></a>
<p>Function names: alignment_length_no_gaps; aln_len_no_gaps; alng <br/>
Command line interface: pk_alignment_length_no_gaps; pk_aln_len_no_gaps; pk_alng</p>
<p>Calculate alignment length excluding sites with gaps.</p>
<p>Longer alignments when excluding sites with gaps is
associated with strong phylogenetic signal.</p>
<p>PhyKIT reports three tab delimited values:
col1: number of sites without gaps
col2: total number of sites
col3: percentage of sites without gaps</p>
<p>Association between alignment length when excluding sites
with gaps and phylogenetic signal was determined by Shen
et al., Genome Biology and Evolution (2016),
doi: 10.1093/gbe/evw179.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>aln_len_no_gaps<span class="w"> </span>&lt;alignment&gt;<span class="w"> </span><span class="o">[</span>--cpu<span class="w"> </span>&lt;cpu&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="alignment-recoding">
<h3>Alignment recoding<a class="headerlink" href="#alignment-recoding" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: alignment_recoding; aln_recoding; recode <br/>
Command line interface: pk_alignment_recoding; pk_aln_recoding; bk_recode</p>
<p>Recode alignments using reduced character states.</p>
<p>Alignments can be recoded using established or
custom recoding schemes. Recoding schemes are
specified using the -c/‚Äìcode argument. Custom
recoding schemes can be used and should be formatted
as a two column file wherein the first column is the
recoded character and the second column is the character
in the alignment.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>alignment_recoding<span class="w"> </span>&lt;fasta&gt;<span class="w"> </span><span class="o">[</span>-c/--code<span class="w"> </span>&lt;code&gt;<span class="w"> </span>--cpu<span class="w"> </span>&lt;cpu&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Codes for which recoding scheme to use: <br/>
<strong>RY-nucleotide</strong> <br/>
R = purines (i.e., A and G) <br/>
Y = pyrimidines (i.e., T and C) <br/></p>
<p><strong>SandR-6</strong> <br/>
0 = A, P, S, and T <br/>
1 = D, E, N, and G <br/>
2 = Q, K, and R <br/>
3 = M, I, V, and L <br/>
4 = W and C <br/>
5 = F, Y, and H <br/></p>
<p><strong>KGB-6</strong> <br/>
0 = A, G, P, and S <br/>
1 = D, E, N, Q, H, K, R, and T <br/>
2 = M, I, and L <br/>
3 = W <br/>
4 = F and Y <br/>
5 = C and V <br/></p>
<p><strong>Dayhoff-6</strong> <br/>
0 = A, G, P, S, and T <br/>
1 = D, E, N, and Q <br/>
2 = H, K, and R <br/>
3 = I, L, M, and V <br/>
4 = F, W, and Y <br/>
5 = C <br/></p>
<p><strong>Dayhoff-9</strong> <br/>
0 = D, E, H, N, and Q <br/>
1 = I, L, M, and V <br/>
2 = F and Y <br/>
3 = A, S, and T <br/>
4 = K and R <br/>
5 = G <br/>
6 = P <br/>
7 = C <br/>
8 = W <br/></p>
<p><strong>Dayhoff-12</strong> <br/>
0 = D, E, and Q <br/>
1 = M, L, I, and V <br/>
2 = F and Y <br/>
3 = K, H, and R <br/>
4 = G <br/>
5 = A <br/>
6 = P <br/>
7 = S <br/>
8 = T <br/>
9 = N <br/>
A = W <br/>
B = C <br/></p>
<p><strong>Dayhoff-15</strong> <br/>
0 = D, E, and Q <br/>
1 = M and L <br/>
2 = I and V <br/>
3 = F and Y <br/>
4 = G <br/>
5 = A <br/>
6 = P <br/>
7 = S <br/>
8 = T <br/>
9 = N <br/>
A = K <br/>
B = H <br/>
C = R <br/>
D = W <br/>
E = C <br/></p>
<p><strong>Dayhoff-18</strong> <br/>
0 = F and Y <br/>
1 = M and L <br/>
2 = I <br/>
3 = V <br/>
4 = G <br/>
5 = A <br/>
6 = P <br/>
7 = S <br/>
8 = T <br/>
9 = D <br/>
A = E <br/>
B = Q <br/>
C = N <br/>
D = K <br/>
E = H <br/>
F = R <br/>
G = W <br/>
H = C <br/></p>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file <br/>
<em>-c/--code</em>: argument to specify the recoding scheme to use <br/>
<em>\-\-cpu</em>: CPUs to use to accelerate calculation</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="column-score">
<h3>Column score<a class="headerlink" href="#column-score" title="Link to this heading">ÔÉÅ</a></h3>
<a class="reference internal image-reference" href="../_images/column_score.png"><img alt="../_images/column_score.png" class="align-center" src="../_images/column_score.png" style="width: 75%;" /></a>
<p>Function names: column_score; cs <br/>
Command line interface: pk_column_score; pk_cs</p>
<p>Calculates column score.</p>
<p>Column is an accuracy metric for a multiple alignment relative
to a reference alignment. It is calculated by summing the correctly
aligned columns over all columns in an alignment. Thus, values range
from 0 to 1 and higher values indicate more accurate alignments.</p>
<p>Column score is calculated following Thompson et al., Nucleic
Acids Research (1999), doi: 10.1093/nar/27.13.2682.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>column_score<span class="w"> </span>&lt;alignment&gt;<span class="w"> </span>--reference<span class="w"> </span>&lt;reference_alignment&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be a query
fasta alignment file to be scored for accuracy <br/>
<em>-r/\-\-reference</em>: reference alignment to compare the query alignment
to</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="compositional-bias-per-site">
<h3>Compositional bias per site<a class="headerlink" href="#compositional-bias-per-site" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: compositional_bias_per_site; comp_bias_per_site; cbps <br/>
Command line interface: pk_compositional_bias_per_site; pk_comp_bias_per_site; pk_cbps</p>
<p>Calculates compositional bias per site in an alignment.</p>
<p>Site-wise chi-squared tests are conducted in an alignment to
detect compositional biases. PhyKIT outputs four columns: <br/>
col 1: index in alignment <br/>
col 2: chi-squared statistic (higher values indicate greater bias) <br/>
col 3: multi-test corrected p-value (Benjamini-Hochberg false discovery rate procedure) <br/>
col 4: uncorrected p-value</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>comp_bias_per_site<span class="w"> </span>&lt;alignment&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be a query
fasta alignment to calculate the site-wise compositional bias of <br/></p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="create-concatenation-matrix">
<h3>Create concatenation matrix<a class="headerlink" href="#create-concatenation-matrix" title="Link to this heading">ÔÉÅ</a></h3>
<a class="reference internal image-reference" href="../_images/create_concat_matrix.png"><img alt="../_images/create_concat_matrix.png" class="align-center" src="../_images/create_concat_matrix.png" style="width: 75%;" /></a>
<p>Function names: create_concatenation_matrix, create_concat, cc <br/>
Command line interface: pk_create_concatenation_matrix, pk_create_concat, pk_cc</p>
<p>Create a concatenated alignment file. This function is
used to help in the construction of multi-locus data
matrices.</p>
<p>PhyKIT will output three files:
1) A fasta file with ‚Äò.fa‚Äô appended to the prefix specified with the -p/\-\-prefix parameter.
2) A partition file ready for input into RAxML or IQ-tree.
3) An occupancy file that summarizes the taxon occupancy per sequence.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>create_concat<span class="w"> </span>-a<span class="w"> </span>&lt;file&gt;<span class="w"> </span>-p<span class="w"> </span>&lt;string&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>-a/\-\-alignment</em>: alignment list file. File should contain a single column list of alignment
sequence files to concatenate into a single matrix. Provide path to files relative to
working directory or provide absolute path. <br/>
<em>-p/\-\-prefix</em>: prefix of output files</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="evolutionary-rate-per-site">
<h3>Evolutionary Rate per Site<a class="headerlink" href="#evolutionary-rate-per-site" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: evolutionary_rate_per_site; evo_rate_per_site; erps <br/>
Command line interface: pk_evolutionary_rate_per_site; pk_evo_rate_per_site; pk_erps</p>
<p>Estimate evolutionary rate per site.</p>
<p>Evolutionary rate per site is one minus the sum of squared frequency of different
characters at a given site. Values may range from 0 (slow evolving; no diversity
at the given site) to 1 (fast evolving; all characters appear only once).</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>evo_rate_per_site<span class="w"> </span>&lt;alignment&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be a query
fasta alignment to calculate the site-wise evolutionary rate of <br/></p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="faidx">
<h3>Faidx<a class="headerlink" href="#faidx" title="Link to this heading">ÔÉÅ</a></h3>
<a class="reference internal image-reference" href="../_images/faidx.png"><img alt="../_images/faidx.png" class="align-center" src="../_images/faidx.png" style="width: 75%;" /></a>
<p>Function names: faidx; get_entry; ge <br/>
Command line interface: pk_faidx; pk_get_entry; pk_ge</p>
<p>Extracts sequence entry from fasta file.</p>
<p>This function works similarly to the faidx function
in samtools, but does not requiring an indexing step.</p>
<p>To obtain multiple entries, input multiple entries separated
by a comma (,). For example, if you want entries
named ‚Äúseq_0‚Äù and ‚Äúseq_1‚Äù, the string ‚Äúseq_0,seq_1‚Äù
should be associated with the -e argument.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>faidx<span class="w"> </span>&lt;fasta&gt;<span class="w"> </span>-e/--entry<span class="w"> </span>&lt;fasta<span class="w"> </span>entry&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-v/\-\-verbose</em>: entry name to be extracted from the inputted fasta file
entry</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="guanine-cytosine-gc-content">
<h3>Guanine-cytosine (GC) content<a class="headerlink" href="#guanine-cytosine-gc-content" title="Link to this heading">ÔÉÅ</a></h3>
<a class="reference internal image-reference" href="../_images/gc_content.png"><img alt="../_images/gc_content.png" class="align-center" src="../_images/gc_content.png" style="width: 75%;" /></a>
<p>Function names: gc_content; gc <br/>
Command line interface: pk_gc_content; pk_gc</p>
<p>Calculate GC content of a fasta file.</p>
<p>GC content is negatively correlated with phylogenetic signal.</p>
<p>If there are multiple entries, use the -v/\-\-verbose option
to determine the GC content of each fasta entry separately.
Association between GC content and phylogenetic signal was
determined by Shen et al., Genome Biology and Evolution (2016),
doi: 10.1093/gbe/evw179.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>gc_content<span class="w"> </span>&lt;fasta&gt;<span class="w"> </span><span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;fasta&gt;</em>: first argument after function name should be a fasta file <br/>
<em>-v/\-\-verbose</em>: optional argument to print the GC content of each fasta
entry</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="pairwise-identity">
<h3>Pairwise identity<a class="headerlink" href="#pairwise-identity" title="Link to this heading">ÔÉÅ</a></h3>
<a class="reference internal image-reference" href="../_images/pairwise_identity.png"><img alt="../_images/pairwise_identity.png" class="align-center" src="../_images/pairwise_identity.png" style="width: 75%;" /></a>
<p>Function names: pairwise_identity; pairwise_id, pi <br/>
Command line interface: pk_pairwise_identity; pk_pairwise_id, pk_pi</p>
<p>Calculate the average pairwise identity among sequences.</p>
<p>Pairwise identities can be used as proxies for the evolutionary rate of sequences.</p>
<p>Pairwise identity is defined as the number of identical
columns (including gaps) between two aligned sequences divided
by the number of columns in the alignment. Summary statistics
are reported unless used with the verbose option in which
all pairwise identities will be reported.</p>
<p>An example of pairwise identities being used as a proxy
for evolutionary rate can be found here: Chen et al.
Genome Biology and Evolution (2017), doi: 10.1093/gbe/evx147.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>pairwise_identity<span class="w"> </span>&lt;alignment&gt;<span class="w"> </span><span class="o">[</span>-v/--verbose<span class="w"> </span>-e/--exclude_gaps<span class="w"> </span>--cpu<span class="w"> </span>&lt;cpu&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file <br/>
<em>-e/\-\-exclude_gaps</em>: if a site has a gap, ignore it <br/>
<em>\-\-cpu</em>: CPUs to use to accelerate calculation</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="parsimony-informative-sites">
<h3>Parsimony informative sites<a class="headerlink" href="#parsimony-informative-sites" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: parsimony_informative_sites; pis <br/>
Command line interface: pk_parsimony_informative_sites; pk_pis</p>
<p>Calculate the number and percentage of parismony
informative sites in an alignment.</p>
<p>The number of parsimony informative sites in an alignment
is associated with strong phylogenetic signal.</p>
<p>PhyKIT reports three tab delimited values:
col1: number of parsimony informative sites
col2: total number of sites
col3: percentage of parsimony informative sites</p>
<p>Association between the number of parsimony informative
sites and phylogenetic signal was determined by Shen
et al., Genome Biology and Evolution (2016),
doi: 10.1093/gbe/evw179 and Steenwyk et al., PLOS Biology
(2020), doi: 10.1371/journal.pbio.3001007.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>parsimony_informative_sites<span class="w"> </span>&lt;alignment&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file <br/>
<em>\-\-cpu</em>: CPUs to use to accelerate calculation</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="protein-to-nucleotide-alignment">
<h3>Protein-to-nucleotide alignment<a class="headerlink" href="#protein-to-nucleotide-alignment" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: thread_dna; pal2nal, p2n <br/>
Command line interface: pk_thread_dna; pk_pal2nal, pk_p2n</p>
<p>Thread DNA sequence onto a protein alignment to create a
codon-based alignment.</p>
<p>This function requires input alignments are in fasta format.
Codon alignments are then printed to stdout. Note, paired
sequences are assumed to have the same name between the
protein and nucleotide file. The order does not matter.</p>
<p>To thread nucleotide sequences over a trimmed amino acid
alignment, provide PhyKIT with a log file specifying which
sites have been trimmed and which have been kept. The log
file must be formatted the same as the log files outputted
by the alignment trimming toolkit ClipKIT (see -l in ClipKIT
documentation.) Details about ClipKIT can be seen here:
<a class="reference external" href="https://github.com/JLSteenwyk/ClipKIT">https://github.com/JLSteenwyk/ClipKIT</a>.</p>
<p>If using a ClipKIT log file, the untrimmed protein alignment
should be provided in the -p/‚Äìprotein argument.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>thread_dna<span class="w"> </span>-p<span class="w"> </span>&lt;file&gt;<span class="w"> </span>-n<span class="w"> </span>&lt;file&gt;<span class="w"> </span><span class="o">[</span>-s<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>-p/\-\-protein</em>: protein alignment file <br/>
<em>-n/\-\-nucleotide</em>: nucleotide sequence file <br/>
<em>-c/\-\-clipkit_log</em>: clipkit outputted log file <br/>
<em>-s/\-\-stop</em>: boolean for whether or not stop codons should be kept.
If used, stop codons will be removed.</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="relative-composition-variability">
<h3>Relative composition variability<a class="headerlink" href="#relative-composition-variability" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: relative_composition_variability; rel_comp_var; rcv <br/>
Command line interface: pk_relative_composition_variability; pk_rel_comp_var; pk_rcv</p>
<p>Calculate RCV (relative composition variability) for an alignment.</p>
<p>Lower RCV values are thought to be desirable because they represent
a lower composition bias in an alignment. Statistically, RCV describes
the average variability in sequence composition among taxa.</p>
<p>RCV is calculated following Phillips and Penny, Molecular Phylogenetics
and Evolution (2003), doi: 10.1016/S1055-7903(03)00057-5.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>relative_composition_variability<span class="w"> </span>&lt;alignment&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file <br/>
<em>\-\-cpu</em>: CPUs to use to accelerate calculation</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="relative-composition-variability-taxon">
<h3>Relative composition variability, taxon<a class="headerlink" href="#relative-composition-variability-taxon" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: relative_composition_variability_taxon; rel_comp_var_taxon; rcvt <br/>
Command line interface: pk_relative_composition_variability_taxon; pk_rel_comp_var_taxon; pk_rcvt</p>
<p>Calculate RCVT (relative composition variability, taxon) for an alignment.</p>
<p>RCVT is the relative composition variability metric for individual taxa.
This facilitates identifying specific taxa that may have compositional
biases. Lower RCVT values are more desirable because they indicate
a lower composition bias for a given taxon in an alignment.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>relative_composition_variability_taxon<span class="w"> </span>&lt;alignment&gt;<span class="w"> </span><span class="o">[</span>--cpu<span class="w"> </span>&lt;cpu&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file <br/>
<em>\-\-cpu</em>: CPUs to use to accelerate calculation</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="rename-fasta-entries">
<h3>Rename FASTA entries<a class="headerlink" href="#rename-fasta-entries" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: rename_fasta_entries; rename_fasta <br/>
Command line interface: pk_rename_fasta_entries; pk_rename_fasta</p>
<p>Renames fasta entries.</p>
<p>Renaming fasta entries will follow the scheme of a tab-delimited
file wherein the first column is the current fasta entry name and
the second column is the new fasta entry name in the resulting
output alignment. Note, the input fasta file does not need to be
an alignment file.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>rename_fasta_entries<span class="w"> </span>&lt;fasta&gt;<span class="w"> </span>-i/--idmap<span class="w"> </span>&lt;idmap&gt;<span class="w"> </span><span class="o">[</span>-o/--output<span class="w"> </span>&lt;output_file&gt;<span class="w"> </span>--cpu<span class="w"> </span>&lt;cpu&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file <br/>
<em>-i/\-\-idmap</em>: identifier map of current FASTA names (col1) and desired FASTA names (col2) <br/>
<em>\-\-cpu</em>: CPUs to use to accelerate calculation</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="sum-of-pairs-score">
<h3>Sum-of-pairs score<a class="headerlink" href="#sum-of-pairs-score" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: sum_of_pairs_score; sops; sop <br/>
Command line interface: pk_sum_of_pairs_score; pk_sops; pk_sop</p>
<p>Calculates sum-of-pairs score.</p>
<p>Sum-of-pairs is an accuracy metric for a multiple alignment relative
to a reference alignment. It is calculated by summing the correctly
aligned residue pairs over all pairs of sequences. Thus, values range
from 0 to 1 and higher values indicate more accurate alignments.</p>
<p>Column score is calculated following Thompson et al., Nucleic
Acids Research (1999), doi: 10.1093/nar/27.13.2682.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>sum_of_pairs_score<span class="w"> </span>&lt;alignment&gt;<span class="w"> </span>--reference<span class="w"> </span>&lt;reference_alignment&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be a query
fasta alignment file to be scored for accuracy <br/>
<em>-r/\-\-reference</em>: reference alignment to compare the query alignment
to</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="variable-sites">
<h3>Variable sites<a class="headerlink" href="#variable-sites" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: variable_sites; vs <br/>
Command line interface: pk_variable_sites; pk_vs</p>
<p>Calculate the number of variable sites in an alignment.</p>
<p>The number of variable sites in an alignment is
associated with strong phylogenetic signal.
PhyKIT reports three tab delimited values:
col1: number of variable sites
col2: total number of sites
col3: percentage of variable sites</p>
<p>Association between the number of variable sites and
phylogenetic signal was determined by Shen et al.,
Genome Biology and Evolution (2016),
doi: 10.1093/gbe/evw179.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>variable_sites<span class="w"> </span>&lt;alignment&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be an alignment file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="tree-based-functions">
<h2>Tree-based functions<a class="headerlink" href="#tree-based-functions" title="Link to this heading">ÔÉÅ</a></h2>
<hr class="docutils" />
<section id="bipartition-support-statistics">
<h3>Bipartition support statistics<a class="headerlink" href="#bipartition-support-statistics" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: bipartition_support_stats; bss <br/>
Command line interface: pk_bipartition_support_stats; pk_bss</p>
<p>Calculate summary statistics for bipartition support.</p>
<p>High bipartition support values are thought to be desirable because
they are indicative of greater certainty in tree topology.</p>
<p>To obtain all bipartition support values, use the -v/\-\-verbose option.
In addition to support values for each node, the names of all terminal
branches tips are also included. Each terminal branch name is separated
with a semi-colon (;).</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>bipartition_support_stats<span class="w"> </span>&lt;tree&gt;<span class="w"> </span><span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be a tree file <br/>
<em>-v/\-\-verbose</em>: optional argument to print all bipartition support values</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="branch-length-multiplier">
<h3>Branch length multiplier<a class="headerlink" href="#branch-length-multiplier" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: branch_length_multiplier; blm <br/>
Command line interface: pk_branch_length_multiplier; pk_blm</p>
<p>Multiply branch lengths in a phylogeny by a given factor.</p>
<p>This can help modify reference trees when conducting simulations
or other analyses.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>branch_length_multiplier<span class="w"> </span>&lt;tree&gt;<span class="w"> </span>-f<span class="w"> </span>n<span class="w"> </span><span class="o">[</span>-o--output<span class="w"> </span>&lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be a tree file <br/>
<em>-f/\-\-factor</em>: factor to multiply branch lengths by <br/>
<em>-o/\-\-output</em>: optional argument to name the outputted tree file. Default
output will have the same name as the input file but with the suffix ‚Äú.factor_(n).tre‚Äù</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="collapse-bipartitions">
<h3>Collapse bipartitions<a class="headerlink" href="#collapse-bipartitions" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: collapse_branches, collapse, cb <br/>
Command line interface: pk_collapse_branches, pk_collapse, pk_cb</p>
<p>Collapse branches on a phylogeny according to bipartition support.</p>
<p>Bipartitions will be collapsed if they are less than the user specified
value.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>collapse_branches<span class="w"> </span>&lt;tree&gt;<span class="w"> </span>-s/--support<span class="w"> </span>n<span class="w"> </span><span class="o">[</span>-o/--output<span class="w"> </span>&lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;alignment&gt;</em>: first argument after function name should be a tree file <br/>
<em>-s/\-\-support</em>: bipartitions with support less than this value will be
collapsed <br/>
<em>-o/\-\-output</em>: optional argument to name the outputted tree file. Default
output will have the same name as the input file but with the suffix
‚Äú.collapsed_(support).tre‚Äù</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="covarying-evolutionary-rates">
<h3>Covarying evolutionary rates<a class="headerlink" href="#covarying-evolutionary-rates" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: covarying_evolutionary_rates; cover <br/>
Command line interface: pk_covarying_evolutionary_rates; pk_cover</p>
<p>Determine if two genes have a signature of covariation with one another.
Genes that have covarying evolutionary histories tend to have
similar functions and expression levels.</p>
<p>Input two phylogenies and calculate the correlation among relative
evolutionary rates between the two phylogenies. The two input trees
do not have to have the same taxa. This function will first prune both
trees to have the same tips. To transform branch lengths into relative
rates, PhyKIT uses the putative species tree‚Äôs branch lengths, which is
inputted by the user. As recommended by the original method developers,
outlier branche lengths are removed. Outlier branches have a relative
evolutionary rate greater than five.</p>
<p>PhyKIT reports two tab delimited values:
col1: correlation coefficient
col2: p-value</p>
<p>Method is empirically evaluated by Clark et al., Genome Research
(2012), doi: 10.1101/gr.132647.111. Normalization method using a
species tree follows Sato et al., Bioinformatics (2005), doi:
10.1093/bioinformatics/bti564.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>covarying_evolutionary_rates<span class="w"> </span>&lt;tree_file_zero&gt;<span class="w"> </span>&lt;tree_file_one&gt;<span class="w"> </span>-r/--reference<span class="w"> </span>&lt;reference_tree_file&gt;<span class="w"> </span><span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree_file_zero&gt;</em>: first argument after function name should be an alignment file <br/>
<em>&lt;tree_file_one&gt;</em>: first argument after function name should be an alignment file <br/>
<em>-r/\-\-reference</em>: a tree to correct branch lengths by in the two input trees. Typically,
this is a putative species tree. <br/>
<em>-v/\-\-verbose</em>: print out corrected branch lengths shared between tree 0 and tree 1</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="degree-of-violation-of-the-molecular-clock">
<h3>Degree of violation of the molecular clock<a class="headerlink" href="#degree-of-violation-of-the-molecular-clock" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: degree_of_violation_of_a_molecular_clock, dvmc <br/>
Command line interface: pk_degree_of_violation_of_a_molecular_clock, pk_dvmc</p>
<p>Calculate degree of violation of a molecular clock (or DVMC) in a phylogeny.</p>
<p>Lower DVMC values are thought to be desirable because they are indicative
of a lower degree of violation in the molecular clock assumption.</p>
<p>Typically, outgroup taxa are not included in molecular clock analysis. Thus,
prior to calculating DVMC from a single gene tree, users may want to prune
outgroup taxa from the phylogeny. To prune tips from a phylogeny, see the
prune_tree function.</p>
<p>Calculate DVMC in a tree following Liu et al., PNAS (2017), doi: 10.1073/pnas.1616744114.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>degree_of_violation_of_a_molecular_clock<span class="w"> </span>&lt;tree&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: input file tree name</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="evolutionary-rate">
<h3>Evolutionary rate<a class="headerlink" href="#evolutionary-rate" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: evolutionary_rate, evo_rate <br/>
Command line interface: pk_evolutionary_rate, pk_evo_rate</p>
<p>Calculate a tree-based estimation of the evolutionary rate of a gene.</p>
<p>Evolutionary rate is the total tree length divided by the number
of terminals.</p>
<p>Calculate evolutionary rate following Telford et al., Proceedings
of the Royal Society B (2014).</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>evolutionary_rate<span class="w"> </span>&lt;tree&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: input file tree name</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="hidden-paralogy-check">
<h3>Hidden paralogy check<a class="headerlink" href="#hidden-paralogy-check" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: hidden_paralogy_check, clan_check <br/>
Command line interface: pk_hidden_paralogy_check, pk_clan_check</p>
<p>Scan tree for evidence of hidden paralogy.</p>
<p>This analysis can be used to identify hidden paralogy.
Specifically, this method will examine if a set of
well known monophyletic taxa are, in fact, monophyletic.
If they are not, the evolutionary history of the gene may
be subject to hidden paralogy. This analysis is typically
done with single-copy orthologous genes.</p>
<p>Requires a clade file, which species which monophyletic
lineages to check for. Multiple monophyletic
lineages can be specified. Each lineage should
be specified on a single line and each tip name
(or taxon name) should be separated by a space.
For example, if it is anticipated that tips
‚ÄúA‚Äù, ‚ÄúB‚Äù, and ‚ÄúC‚Äù are monophyletic and ‚ÄúD‚Äù,
‚ÄúE‚Äù, and ‚ÄúF‚Äù are expected to be monophyletic, the
clade file should be formatted as follows: <br/>
‚Äú <br/>
A B C <br/>
D E F <br/>
‚Äú</p>
<p>The output will report if the specified taxa were monophyletic
or not. The number of rows will reflect how many groups of taxa
were checked for monophyly. For example,
if there were three rows of clades in the -c file, there will be
three rows in the output
where the first row in the output corresponds to the
results of the first row in the clade file. <br/></p>
<p>The concept behind this analysis follows
Siu-Ting et al., Molecular Biology and Evolution (2019),
doi: 10.1093/molbev/msz067.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>hidden_paralogy_check<span class="w"> </span>&lt;tree&gt;<span class="w"> </span>-c/--clade<span class="w"> </span>&lt;clade_file&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>-t/\-\-tree</em>: input file tree name
<em>-c/\-\-clade</em>: clade file detailing which monophyletic lineages should
be scanned for</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="internal-branch-statistics">
<h3>Internal branch statistics<a class="headerlink" href="#internal-branch-statistics" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: internal_branch_stats; ibs <br/>
Command line interface: pk_internal_branch_stats; pk_ibs</p>
<p>Calculate summary statistics for internal branch lengths in a phylogeny.</p>
<p>Internal branch lengths can be useful for phylogeny diagnostics.</p>
<p>To obtain all internal branch lengths, use the -v/\-\-verbose option.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>internal_branch_stats<span class="w"> </span>&lt;tree&gt;<span class="w"> </span><span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>-v/\-\-verbose</em>: optional argument to print all internal branch lengths</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="internode-labeler">
<h3>Internode labeler<a class="headerlink" href="#internode-labeler" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: internode_labeler; il <br/>
Command line interface: pk_internode_labeler; pk_il</p>
<p>Appends numerical identifiers to bipartitions in place of support values.
This is helpful for pointing to specific internodes in supplementary files
or otherwise.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>internode_labeler<span class="w"> </span>&lt;tree&gt;<span class="w"> </span><span class="o">[</span>-o/--output<span class="w"> </span>&lt;file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>-o/\-\-output</em>: optional argument to name the outputted tree file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="last-common-ancestor-subtree">
<h3>Last common ancestor subtree<a class="headerlink" href="#last-common-ancestor-subtree" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: last_common_ancestor_subtree; lca_subtree <br/>
Command line interface: pk_last_common_ancestor_subtree; pk_lca_subtree</p>
<p>Obtains subtree from a phylogeny by getting the last common ancestor
from a list of taxa.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>last_common_ancestor_subtree<span class="w"> </span>&lt;file&gt;<span class="w"> </span>&lt;list_of_taxa&gt;<span class="w"> </span><span class="o">[</span>-o/--output<span class="w"> </span>&lt;file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>&lt;list_of_taxa&gt;</em>: second argument after function name should be a single column
file with the list of taxa to get the last common ancestor subtree for
<em>-o/\-\-output</em>: optional argument to print all LB score values</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="long-branch-score">
<h3>Long branch score<a class="headerlink" href="#long-branch-score" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: long_branch_score; lb_score; lbs <br/>
Command line interface: pk_long_branch_score; pk_lb_score; pk_lbs</p>
<p>Calculate long branch (LB) scores in a phylogeny.</p>
<p>Lower LB scores are thought to be desirable because
they are indicative of taxa or trees that likely do
not have issues with long branch attraction.</p>
<p>LB score is the mean pairwise patristic distance of
taxon i compared to all other taxa over the average
pairwise patristic distance.</p>
<p>PhyKIT reports summary statistics. To obtain LB scores
for each taxa, use the -v/‚Äìverbose option.</p>
<p>LB scores are calculated following Struck, Evolutionary
Bioinformatics (2014), doi: 10.4137/EBO.S14239.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>long_branch_score<span class="w"> </span>&lt;tree&gt;<span class="w"> </span><span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>-v/\-\-verbose</em>: optional argument to print all LB score values</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="monophyly-check">
<h3>Monophyly check<a class="headerlink" href="#monophyly-check" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: monophyly_check; is_monophyletic <br/>
Command line interface: pk_monophyly_check; pk_is_monophyletic</p>
<p>This analysis can be used to determine if a set of
taxa are exclusively monophyletic. By exclusively monophyletic,
if other taxa are in the same clade, the lineage will not be
considered exclusively monophyletic.</p>
<p>Requires a taxa file, which species which tip names
are expected to be monophyletic. File format is a
single column file with tip names. Tip names not
present in the tree will not be considered when
examining monophyly.</p>
<p>The output will have six columns.
col 1: if the clade was or wasn‚Äôt monophyletic
col 2: average bipartition support value in the clade of interest
col 3: maximum bipartition support value in the clade of interest
col 4: minimum bipartition support value in the clade of interest
col 5: standard deviation of bipartition support values in the clade of interest
col 6: tip names of taxa monophyletic with the lineage of interest excluding those that are listed in the taxa_of_interest file</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>monophyly_check<span class="w"> </span>&lt;tree&gt;<span class="w"> </span>&lt;list_of_taxa&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>&lt;list_of_taxa&gt;</em>: single column file with list of tip names to
examine the monophyly of</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="nearest-neighbor-interchange">
<h3>Nearest neighbor interchange<a class="headerlink" href="#nearest-neighbor-interchange" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: nearest_neighbor_interchange; nni <br/>
Command line interface: pk_nearest_neighbor_interchange; pk_nni</p>
<p>Generate all nearest neighbor interchange moves for a binary
rooted tree.</p>
<p>By default, the output file will have the same name as the input
file but with the suffix ‚Äú.nnis‚Äù</p>
<p>The output file will also include the original phylogeny.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>nearest_neighbor_interchange<span class="w"> </span>&lt;tree&gt;<span class="w"> </span><span class="o">[</span>-o/--output<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>-o/\-\-output</em>: optional argument to specify output file name</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="patristic-distances">
<h3>Patristic distances<a class="headerlink" href="#patristic-distances" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: patristic_distances; pd <br/>
Command line interface: pk_patristic_distances; pk_pd</p>
<p>Calculate summary statistics among patristic distances in a phylogeny.</p>
<p>Patristic distances are all tip-to-tip distances in a phylogeny.</p>
<p>To obtain all patristic distances, use the -v/‚Äìverbose option.
With the -v option, the first column will have two taxon names
separated by a ‚Äò-‚Äô followed by the patristic distance. Features
will be tab separated.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>patristic_distances<span class="w"> </span>&lt;tree&gt;<span class="w"> </span><span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>-o/\-\-output</em>: optional argument to print all tip-to-tip distances</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="polytomy-testing">
<h3>Polytomy testing<a class="headerlink" href="#polytomy-testing" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: polytomy_test; polyt_test; polyt; ptt <br/>
Command line interface: pk_polytomy_test; pk_polyt_test; pk_polyt; pk_ptt</p>
<p>Conduct a polytomy test for three clades in a phylogeny.</p>
<p>Polytomy tests can be used to identify putative radiations
as well as identify well supported alternative topologies.</p>
<p>The polytomy testing function takes as input a file with
the three groups of taxa to test the relationships for and
a single column file with the names of the desired tree files
to use for polytomy testing. Next, the script to examine
support for the grouping of the three taxa using triplets
and gene support frequencies.</p>
<p>This function can account for uncertainty in gene trees -
that is, the input phylogenies can have collapsed bipartitions.</p>
<p>Thereafter, a chi-squared test is conducted to determine if there
is evidence to reject the null hypothesis wherein the null
hypothesis is that the three possible topologies among the three
groups are equally supported. This test is done using gene support
frequencies.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>polytomy_test<span class="w"> </span>-t/--trees<span class="w"> </span>&lt;trees&gt;<span class="w"> </span>-g/--groups<span class="w"> </span>&lt;groups&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>-t/\-\-trees &lt;trees&gt;</em>: single column file with the names of
phylogenies to use for polytomy testing <br/>
<em>-g/\-\-groups</em>: a tab-delimited file with the grouping designations
to test. Lines starting with commetns are not considered. Names of
individual taxa should be separated by a semi-colon ‚Äò;‚Äô</p>
<p>For example, the groups file could look like the following:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span><span class="c1">#label group0  group1  group2</span>
name_of_test<span class="w">    </span>tip_name_A<span class="p">;</span>tip_name_B<span class="w">   </span>tip_name_C<span class="w">  </span>tip_name_D<span class="p">;</span>tip_name_E
</pre></div>
</div>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="print-tree">
<h3>Print tree<a class="headerlink" href="#print-tree" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: print_tree; print; pt <br/>
Command line interface: pk_print_tree; pk_print; pk_pt</p>
<p>Print ascii tree of input phylogeny.</p>
<p>Phylogeny can be printed with or without branch lengths.
By default, the phylogeny will be printed with branch lengths
but branch lengths can be removed using the -r/‚Äìremove argument.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>print_tree<span class="w"> </span>&lt;tree&gt;<span class="w"> </span><span class="o">[</span>-r/--remove<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>-r/\-\-remove</em>: optional argument to print the phylogeny without branch
lengths</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="prune-tree">
<h3>Prune tree<a class="headerlink" href="#prune-tree" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: prune_tree; prune <br/>
Command line interface: pk_prune_tree; pk_prune</p>
<p>Prune tips from a phylogeny.</p>
<p>Provide a single column file with the names of the tips
in the input phylogeny you would like to prune from the
tree.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>prune_tree<span class="w"> </span>&lt;tree&gt;<span class="w"> </span>&lt;list_of_taxa&gt;<span class="w"> </span><span class="o">[</span>-o/--output<span class="w"> </span>&lt;output_file&gt;
-k/--keep<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>&lt;list_of_taxa&gt;</em>: single column file with the names of the tips to remove
from the phylogeny <br/>
<em>-o/\-\-output</em>: name of output file for the pruned phylogeny.
Default output will have the same name as the input file but with the suffix
‚Äú.pruned‚Äù
<em>-k/‚Äìkeep</em>: optional argument. If used instead of pruning taxa in &lt;list_of_taxa&gt;,
keep them
|</p>
</section>
<section id="rename-tree-tips">
<h3>Rename tree tips<a class="headerlink" href="#rename-tree-tips" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: rename_tree; rename_tips <br/>
Command line interface: pk_rename_tree; pk_rename_tips</p>
<p>Renames tips in a phylogeny.</p>
<p>Renaming tip files will follow the scheme of a tab-delimited
file wherein the first column is the current tip name and the
second column is the desired tip name in the resulting
phylogeny.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>rename_tree_tips<span class="w"> </span>&lt;tree&gt;<span class="w"> </span>-i/--idmap<span class="w"> </span>&lt;idmap.txt&gt;<span class="w"> </span><span class="o">[</span>-o/--output<span class="w"> </span>&lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>-i/\-\-idmap</em>: identifier map of current tip names (col1) and desired
tip names (col2) <br/>
<em>-o/\-\-output</em>: optional argument to write the renamed tree files to. Default
output will have the same name as the input file but with the suffix ‚Äú.renamed‚Äù</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="robinson-foulds-distance">
<h3>Robinson-Foulds distance<a class="headerlink" href="#robinson-foulds-distance" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: robinson_foulds_distance; rf_distance; rf_dist; rf <br/>
Command line interface: pk_robinson_foulds_distance; pk_rf_distance; pk_rf_dist; pk_rf</p>
<p>Calculate Robinson-Foulds (RF) distance between two trees.</p>
<p>Low RF distances reflect greater similarity between two phylogenies.
This function prints out two values, the plain RF value and the
normalized RF value, which are separated by a tab. Normalized RF values
are calculated by taking the plain RF value and dividing it by 2(n-3)
where n is the number of tips in the phylogeny. Prior to calculating
an RF value, PhyKIT will first determine the number of shared tips
between the two input phylogenies and prune them to a common set of
tips. Thus, users can input trees with different topologies and
infer an RF value among subtrees with shared tips.</p>
<p>PhyKIT will print out
col 1; the plain RF distance and
col 2: the normalized RF distance.</p>
<p>RF distances are calculated following Robinson &amp; Foulds, Mathematical
Biosciences (1981), doi: 10.1016/0025-5564(81)90043-2.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>robinson_foulds_distance<span class="w"> </span>&lt;tree_file_zero&gt;<span class="w"> </span>&lt;tree_file_one&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree_file_zero&gt;</em>: first argument after function name should be a tree file
<em>&lt;tree_file_one&gt;</em>: second argument after function name should be a tree file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="root-tree">
<h3>Root tree<a class="headerlink" href="#root-tree" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: root_tree; root; rt <br/>
Command line interface: pk_root_tree; pk_root; pk_rt</p>
<p>Roots phylogeny using user-specified taxa.</p>
<p>A list of taxa to root the phylogeny on should be specified using the -r
argument. The root_taxa file should be a single-column file with taxa names.
The outputted file will have the same name as the inputted tree file but with
the suffix ‚Äú.rooted‚Äù.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>root_tree<span class="w"> </span>&lt;tree&gt;<span class="w"> </span>-r/--root<span class="w"> </span>&lt;root_taxa&gt;<span class="w"> </span><span class="o">[</span>-o/--output<span class="w"> </span>&lt;output_file&gt;<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file to root|br|
<em>-r/\-\-root</em>: single column file with taxa names to root the phylogeny on|br|
<em>-o/\-\-output</em>: optional argument to specify the name of the output file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="spurious-homolog-identification">
<h3>Spurious homolog identification<a class="headerlink" href="#spurious-homolog-identification" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: spurious_sequence; spurious_seq; ss <br/>
Command line interface: pk_spurious_sequence; pk_spurious_seq; pk_ss</p>
<p>Determines potentially spurious homologs using branch lengths.</p>
<p>Identifies potentially spurious sequences and reports
tips in the phylogeny that could possibly be removed
from the associated multiple sequence alignment. PhyKIT
does so by identifying and reporting long terminal branches
defined as branches that are equal to or 20 times the median
length of all branches.</p>
<p>PhyKIT reports the following information
col1: name of tip that is a putatively spurious sequence
col2: length of branch leading to putatively spurious sequence
col3: threshold used to identify putatively spurious sequences
col4: median branch length in the phylogeny</p>
<p>If there are no putatively spurious sequences, ‚ÄúNone‚Äù is reported.</p>
<p>Using this method to identify potentially spurious sequences
was, to my knowledge, first introduced by Shen et al., (2018)
Cell doi: 10.1016/j.cell.2018.10.023.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>spurious_seq<span class="w"> </span>&lt;file&gt;<span class="w"> </span>-f/<span class="se">\\</span>-<span class="se">\\</span>-factor
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;file&gt;</em>: first argument after function name should be a tree file
<em>-f/\-\-factor</em>: factor to multiply median branch length by to calculate
the threshold of long branches. (Default: 20)</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="terminal-branch-statistics">
<h3>Terminal branch statistics<a class="headerlink" href="#terminal-branch-statistics" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: terminal_branch_stats; tbs <br/>
Command line interface: pk_terminal_branch_stats; pk_tbs</p>
<p>Calculate summary statistics for terminal branch lengths in a phylogeny.</p>
<p>Terminal branch lengths can be useful for phylogeny diagnostics.</p>
<p>To obtain all terminal branch lengths, use the -v/\-\-verbose option.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>terminal_branch_stats<span class="w"> </span>&lt;tree&gt;<span class="w"> </span><span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file <br/>
<em>-v/\-\-verbose</em>: optional argument to print all terminal branch lengths</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="tip-labels">
<h3>Tip labels<a class="headerlink" href="#tip-labels" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: tip_labels; tree_labels; labels; tl <br/>
Command line interface: pk_tip_labels; pk_tree_labels; pk_labels; pk_tl</p>
<p>Prints the tip labels (or names) a phylogeny.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>tip_labels<span class="w"> </span>&lt;tree&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="tip-to-tip-distance">
<h3>Tip-to-tip distance<a class="headerlink" href="#tip-to-tip-distance" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: tip_to_tip_distance; t2t_dist; t2t <br/>
Command line interface: pk_tip_to_tip_distance; pk_t2t_dist; pk_t2t</p>
<p>Calculate distance between two tips (or leaves) in a phylogeny.</p>
<p>Distances are in substitutions per site.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>tip_to_tip_distance<span class="w"> </span>&lt;tree_file&gt;<span class="w"> </span>&lt;tip_1&gt;<span class="w"> </span>&lt;tip_2&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree_file&gt;</em>: first argument after function name should be a tree file <br/>
<em>&lt;tip_1&gt;</em>: second argument should be the name of the first tip of interest <br/>
<em>&lt;tip_2&gt;</em>: third argument should be the name of the second tip of interest</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="tip-to-tip-node-distance">
<h3>Tip-to-tip node distance<a class="headerlink" href="#tip-to-tip-node-distance" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: tip_to_tip_node_distance; t2t_node_dist; t2t_nd <br/>
Command line interface: pk_tip_to_tip_node_distance; pk_t2t_node_dist; pk_t2t_nd</p>
<p>Calculate distance between two tips (or leaves) in a phylogeny.</p>
<p>Distance is measured by the number of nodes between one tip
and another.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>tip_to_tip_node_distance<span class="w"> </span>&lt;tree_file&gt;<span class="w"> </span>&lt;tip_1&gt;<span class="w"> </span>&lt;tip_2&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree_file&gt;</em>: first argument after function name should be a tree file <br/>
<em>&lt;tip_1&gt;</em>: second argument should be the name of the first tip of interest <br/>
<em>&lt;tip_2&gt;</em>: third argument should be the name of the second tip of interest</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="total-tree-length">
<h3>Total tree length<a class="headerlink" href="#total-tree-length" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: total_tree_length; tree_len <br/>
Command line interface: pk_total_tree_length; pk_tree_len</p>
<p>Calculate total tree length, which is a sum of all branches.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>total_tree_length<span class="w"> </span>&lt;tree&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="treeness">
<h3>Treeness<a class="headerlink" href="#treeness" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: treeness; tness <br/>
Command line interface: pk_treeness; pk_tness</p>
<p>Calculate treeness statistic for a phylogeny.</p>
<p>Higher treeness values are thought to be desirable because they
represent a higher signal-to-noise ratio.</p>
<p>Treeness is the sum of internal branch lengths divided by the total
tree length. Therefore, values range from 0 to 1. Treeness can be
used as a measure of the signal-to-noise ratio in a phylogeny.</p>
<p>Calculate treeness (also referred to as stemminess) following
Lanyon, The Auk (1988), doi: 10.1093/auk/105.3.565 and
Phillips and Penny, Molecular Phylogenetics and Evolution
(2003), doi: 10.1016/S1055-7903(03)00057-5.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>treeness<span class="w"> </span>&lt;tree&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>&lt;tree&gt;</em>: first argument after function name should be a tree file</p>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="alignment-and-tree-based-functions">
<h2>Alignment- and tree-based functions<a class="headerlink" href="#alignment-and-tree-based-functions" title="Link to this heading">ÔÉÅ</a></h2>
<hr class="docutils" />
<section id="saturation">
<h3>Saturation<a class="headerlink" href="#saturation" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: saturation; sat <br/>
Command line interface: pk_saturation; pk_sat</p>
<p>Calculate saturation for a given tree and alignment.</p>
<p>Saturation is defined as sequences in multiple sequence
alignments that have undergone numerous substitutions such
that the distances between taxa are underestimated.</p>
<p>Data with no saturation will have a value of 1. The closer
the value is to 1, the less saturated the data.</p>
<p>This function outputs two values (as of v1.19.9). The first
value is the saturation value and the second column is the absolute
value of saturation minus 1. Thus, lower values in the second column
are indicative of values closer to one and, thus, less saturation.</p>
<p>Saturation is calculated following Philippe et al., PLoS
Biology (2011), doi: 10.1371/journal.pbio.1000602.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>saturation<span class="w"> </span>-a<span class="w"> </span>&lt;alignment&gt;<span class="w"> </span>-t<span class="w"> </span>&lt;tree&gt;<span class="w"> </span><span class="o">[</span>-v/--verbose<span class="o">]</span>
</pre></div>
</div>
<p>Options: <br/>
<em>-a/\-\-alignment</em>: an alignment file <br/>
<em>-t/\-\-tree</em>: a tree file <br/>
<em>-v/\-\-verbose</em>: print out patristic distances and uncorrected <br/>
distances used to determine saturation</p>
</section>
<section id="treeness-over-rcv">
<h3>Treeness over RCV<a class="headerlink" href="#treeness-over-rcv" title="Link to this heading">ÔÉÅ</a></h3>
<p>Function names: treeness_over_rcv; toverr; tor <br/>
Command line interface: pk_treeness_over_rcv; pk_toverr; pk_tor</p>
<p>Calculate treeness/RCV for a given alignment and tree.</p>
<p>Higher treeness/RCV values are thought to be desirable because
they harbor a high signal-to-noise ratio are least susceptible
to composition bias.</p>
<p>PhyKIT reports three tab delimited values:
col1: treeness/RCV
col2: treeness
col3: RCV</p>
<p>Calculate treeness/RCV following Phillips and Penny, Molecular
Phylogenetics and Evolution (2003), doi: 10.1016/S1055-7903(03)00057-5.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>phykit<span class="w"> </span>treeness_over_rcv<span class="w"> </span>-a/--alignment<span class="w"> </span>&lt;alignment&gt;<span class="w"> </span>-t/--tree<span class="w"> </span>&lt;tree&gt;
</pre></div>
</div>
<p>Options: <br/>
<em>-a/\-\-alignment</em>: an alignment file <br/>
<em>-t/\-\-tree</em>: a tree file</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../about/index.html" class="btn btn-neutral float-left" title="About" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../tutorials/index.html" class="btn btn-neutral float-right" title="Tutorials" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020 Jacob L. Steenwyk.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-104875636-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-104875636-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>