<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Change log &mdash; phykit  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=b745eb55" />

  
    <link rel="shortcut icon" href="../_static/flavicon.png"/>
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Other software" href="../other_software/index.html" />
    <link rel="prev" title="Tutorials" href="../tutorials/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../about/index.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="../other_software/index.html">Other software</a></li>
<li class="toctree-l1"><a class="reference internal" href="../frequently_asked_questions/index.html">FAQ</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">phykit</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Change log</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="change-log">
<span id="id1"></span><h1>Change log<a class="headerlink" href="#change-log" title="Link to this heading"></a></h1>
<hr class="docutils" />
<p>Major changes to PhyKIT are summarized here.</p>
<p><strong>1.19.4</strong>:
Saturation function forces y-intercept to be zero when calculating slope</p>
<p><strong>1.19.3</strong>:
Saturation function now uses uncorrected distances instead of pairwise identities</p>
<p><strong>1.19.2</strong>:
Verbose pairwise identity reporting separates pairwise identities by tabs and not a dash</p>
<p><strong>1.19.0</strong>:
Added function to test for site-wise compositional biases in an alignment. See function compositional_bias_per_site.</p>
<p><strong>1.18.0</strong>:
Added function to estimate site-wise evolutionary rate in an alignment. See function evolutionary_rate_per_site.</p>
<p><strong>1.15.0</strong>:
Added function to recode alignments based on 8 different recoding schemes (7 for amino acids;
1 for nucleotides). See function recode.</p>
<p><strong>1.14.0</strong>:
Added an optional argument to the thread_dna function. Now, PhyKIT can thread nucleotide
sequences onto a trimmed amino acid alignment. To do so, point PhyKIT to the ClipKIT outputted log
file using the -c argument. The ClipKIT log file can be generated when trimming an alignment with
ClipKIT by adding the -l argument (see here for more details: <a class="reference external" href="https://jlsteenwyk.com/ClipKIT/">https://jlsteenwyk.com/ClipKIT/</a>).</p>
<p><strong>1.12.6</strong>: relative composition variability is now adapted for calculating compositional biases in
individual taxa. The new function in rcvt (relative composition variability, taxon).</p>
<p><strong>1.12.4</strong>: calculations of pairwise identity in alignment now supports excluding pairwise
combinations with gaps.</p>
<p><strong>1.12.3</strong>: hidden paralogy check now simply looks for monophyly or lack thereof for a set of taxa. Hidden paralogy
check still reports insufficient taxon representation.</p>
<p><strong>1.12.2</strong>: removed root.txt file from DVMC function. User’s are now recommended to trim outgroup taxa beforehand</p>
<p><strong>1.11.3</strong>: Added an optional argument to the prune_tree function wherein instead of pruning tips
specified in the input file, those tips will be kept.</p>
<p><strong>1.11.1</strong>: Modified sum of pairs score to divide the correct number
of pairs by the number of pairs in the reference alignment rather
than the query alignment alignment</p>
<p><strong>1.11.0</strong>: Added terminal_branch_stats (alias: tbs) function to examine terminal branch lengths</p>
<p><strong>1.10.1</strong>: Modified column score and sum of pairs score to divide the correct number
of columns or pairs by the number of columns or pairs in the query alignment rather
than the reference alignment</p>
<p><strong>1.10.0</strong>: Added tip_to_tip_node_distance (alias: t2t_node_dist; t2t_nd) function to calculate
the phylogenetic distance between two leaves in a phylogeny. Distance is measured in nodes between
two leaves</p>
<p><strong>1.9.0</strong>: Added monophyly_check (alias: is_monophyletic) function to examine monophyly
among a specified set of taxa</p>
<p><strong>1.8.0</strong>: Added hidden_paralogy_check (alias: clan_check) function to examine phylogenetic
tree for issues of hidden paralogy</p>
<p><strong>1.7.0</strong>: Added nearest_neighbor_interchange (alias: nni) function to generate all NNI moves
for a binary rooted phylogeny</p>
<p><strong>1.6.0</strong>: Added tip_to_tip_distance (alias: t2t_dist; t2t) function to calculate phylogenetic distance
between two leaves in a phylogeny</p>
<p><strong>1.5.0</strong>: Added root_tree (alias: root; rt) function to root a phylogenetic tree</p>
<p><strong>1.4.0</strong>: PhyKIT is now Python version 3.9 and BioPython 1.79 compatible</p>
<p><strong>1.3.0</strong>: Added function that estimates the evolutionary rate of a gene using tree-based
properties. Function name is ‘evolutionary_rate’ or ‘evo_rate’</p>
<p><strong>1.2.2</strong>: added function to get the subtree of the last common ancestor among a set of taxa</p>
<p><strong>1.2.0</strong>: added command line interfaces for all functions so that each command
can easily be executed. For example, ‘phykit aln_len -h’ can now be
called using ‘pk_aln_len -h’</p>
<p><strong>1.1.0</strong>: added faidx (alias: get_entry; ge) function to extract fasta entries from a
multi-fasta file</p>
<p><strong>1.0.3</strong>: added rooting procedure before calculating RF to handle comparing unrooted
and rooted trees</p>
<p><strong>1.0.2</strong>: function that calculates Robinson Foulds distance (robinson_foulds_distance;
rf_distance; rf_dist; rf) now can take trees that differ in topology. PhyKIT
will first determine shared tips between the two trees and prune both trees
to a common set of tips. Next, PhyKIT will calculate the Robinson Foulds
distance.</p>
<p><strong>0.1.3</strong>: Added function (column_score; cs) to calculate the quality of
an alignment given an input query alignment and a reference
alignment to compare it to</p>
<p><strong>0.1.2</strong>: Added function (sum_of_pairs_score; sops; sop) to calculate
the quality of an alignment given an input query alignment
and a reference alignment to compare it to</p>
<p><strong>0.0.9</strong>: PhyKIT now handles error stemming from piping output</p>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../tutorials/index.html" class="btn btn-neutral float-left" title="Tutorials" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../other_software/index.html" class="btn btn-neutral float-right" title="Other software" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020 Jacob L. Steenwyk.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-104875636-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-104875636-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>